// encode-state-machine/index.js
const XState = require('xstate')
const { Machine } = XState

module.exports = Machine({
  initial: 'idle',
  states: {
    idle: {
      on: {
        EVENT_HOSTING_START: {
          target: 'connect_to_attestor'
        },
        EVENT_HOSTING_REPAIR: { // new_hosting_setup
          target: 'hosting_repair'
        },
      },
    },
    connect_to_attestor: {
      actions: ['p2plex_to_attestor'],
      on: {
        _RESOLVE: 'encode_data_and_send_to_attestor',
        _REJECT: 'failure_connect'  // retry
      }
    },
    encode_data_and_send_to_attestor: {
      actions: ['get_and_verify_original_data', 'encode', 'send_data_to_attestor'], // streaming, all happening in parallel
      exit: ['disconnect'],
      on: {
        _RESOLVE: 'idle',
        _REJECT: 'failure_connect'  // retry
      }
    },
    hosting_repair: {
      actions: ['resume'], // streaming, all happening in parallel
      exit: ['disconnect'],
      on: {
        _RESOLVE: 'idle',
        _REJECT: 'failure_resume'  // retry
      }
    },
    failure_connect: {
      on: {
        RETRY: 'connect_to_attestor'
      }
    },
    failure_resume: {
      on: {
        RETRY: 'hosting_repair'
      }
    }
  }
})
