// user-sm/index.js
const XState = require('xstate')
const { Machine, interpret, assign } = XState

module.exports = Machine({
  initial: 'idle',
  states: {
    idle: {
      on: {
        _ACCOUNT_CREATE: {
          target: 'create'
        },
        _ACCOUNT_DELETE: {
          target: 'delete'
        },
        _ACCOUNT_UPDATE: {
          target: 'update'
        }
      },
    },
    create: {
      entry: ['make_or_import_chain_keypair', 'init_nonce'],
      on: {
        _RESOLVE: 'roles_register',
        _REJECT: 'fail_create'
      }
    },
    roles_register: {
      entry: ['make_noise_keypair', 'init_storage', 'make_hyperlog'],
      on: {
        _RESOLVE: 'submit_to_chain',
        _REJECT: 'fail_roles_register'
      }
    },
    update: {
      entry: ['update'],
      on: {
        RESOLVE: 'submit_to_chain',
        REJECT: 'fail_update'
      }
    },
    delete: {
      entry: ['delete'],
      on: {
        RESOLVE: 'submit_to_chain',
        REJECT: 'fail_delete'
      }
    },
    submit_to_chain: {
      entry: ['submit'],
      on: {
        RESOLVE: 'idle',
        REJECT: 'fail_submit_to_chain'
      }
    },
    fail_create: {
      on: {
        _RETRY: 'create'
      }
    },
    fail_update: {
      on: {
        _RETRY: 'update'
      }
    },
    fail_delete: {
      on: {
        _RETRY: 'delete'
      }
    },
    fail_roles_register: {
      on: {
        _RETRY: 'roles_register'
      }
    },
    fail_submit_to_chain: {
      on: {
        _RETRY: 'submit_to_chain'
      }
    },
  }
})
