// attest-state-machine/index.js
const performance_benchmark_lead = require('./sub-performance-benchmark-lead')
const performance_benchmark_regular = require('./sub-performance-benchmark-regular')
const hosting_start = require('./sub-hosting-start')
const proof_of_storage = require('./sub-proof-of-storage')
const XState = require('xstate')
const { Machine } = XState

module.exports = Machine({
  initial: 'idle',
  context: {
   fresh: true,
   restart: false,
   restart_as_new_attestor: false
 },
  states: {
    idle: {
      id: 'idleState',
      on: {
        EVENT_HOSTING_START: {
          target: 'hosting_start'
        },
        EVENT_PERFORMANCE_BENCHMARK_LEAD: {
          target: 'performance_benchmark_lead'
        },
        EVENT_PERFORMANCE_BENCHMARK_REGULAR: {
          target: 'performance_benchmark_regular'
        },
        EVENT_PROOF_OF_STORAGE_REQUEST: {
          target: 'proof_of_storage_request'
        },
      },
    },

    hosting_start: {
      on: {},
      ...hosting_start
    },

    proof_of_storage_request: {
      on: {},
      ...proof_of_storage
    },
    performance_benchmark_regular: {
      on: {},
      ...performance_benchmark_regular
    },
    performance_benchmark_lead: {
      on: {},
      ...performance_benchmark_lead
    },

  }
})
