// attest-state-machine/sub-performance-benchmark-lead.js
module.exports = {
  initial:'p2plex_to_hosters',
  states: {
    p2plex_to_hosters: {
      entry: ['p2plex_connect_to_all_hosters'],
      on: {
        RESOLVE: 'join_attestors_swarm',
        REJECT: 'fail_p2plex_to_hosters',
      }
    },
    join_attestors_swarm: {
      entry: ['join_attestors_swarm'],
      on: {
        RESOLVE: 'idle',
        REJECT: 'fail_join_attestors_swarm'
      }
    },
    idle: {
      on: {
        HOSTER_REPORT: {
          target: 'receive_hoster_report'
        },
        SEND_REPORT: {
          target: 'send_report'
        },
      }
    },
    receive_hoster_report: {
      entry: ['receive_report'],
      on: {
        RESOLVE: 'compare_reports',
        REJECT: 'fail_receive_hoster_report'
      }
    },
    compare_reports: {
      entry: ['compare_reports_with_other_attestors'],
      on: {
        RESOLVE: 'store_results',
        REJECT: 'fail_compare_reports'
      }
    },
    store_results: {
      entry: ['store_results_to_hyperlog'],
      on: {
        RESOLVE: 'organize_performance_check',
        REJECT: 'fail_store_results'
      }
    },
    organize_performance_check: {
      entry: ['select_attestors', 'order_chunks_retrieval'],
      on: {
        RESOLVE: 'get_attestors_reports',
        REJECT: 'fail_organize_performance_check'
      }
    },
    get_attestors_reports: {
      entry: ['get_reports'],
      on: {
        RESOLVE: 'idle',
        REJECT: 'fail_get_attestors_reports'
      }
    },
    send_report: {
      entry: ['prepare_performance_benchmark_report', 'submit_to_chain'],
      exit: ['disconnect'],
      on: {
        RESOLVE: '#idleState',
        REJECT: 'fail_send_report'
      }
    },
    fail_p2plex_to_hosters: {
      on: {
        RETRY: 'p2plex_to_hosters'
      }
    },
    fail_join_attestors_swarm: {
      on: {
        RETRY: 'join_attestors_swarm'
      }
    },
    fail_receive_hoster_report: {
      on: {
        RETRY: 'receive_hoster_report'
      }
    },
    fail_compare_reports: {
      on: {
        RETRY: 'compare_reports'
      }
    },
    fail_store_results: {
      on: {
        RETRY: 'store_results'
      }
    },
    fail_organize_performance_check: {
      on: {
        RETRY: 'organize_performance_check'
      }
    },
    fail_get_attestors_reports: {
      on: {
        RETRY: 'get_attestors_reports'
      }
    },
    fail_send_report: {
      on: {
        RETRY: 'send_report'
      }
    },
  }
}
