// attest-state-machine/sub-hosting-start.js
module.exports = {
  initial: 'connect_to_encoders_and_hosters',
  states: {
    connect_to_encoders_and_hosters: {
      entry: ['connect_to_encoders', 'connect_to_hosters'],
      on: {
       _RESOLVE: 'data_from_encoders_to_hosters',
       _REJECT: 'fail_to_connect' // track who failed, to know who should reconnect in retry
      }
    },
    data_from_encoders_to_hosters: {
      entry: ['get_encoded_compare_and_forward_to_hosters'],
      exit: ['disconnect'],
      on: {
        _RESOLVE: 'report_to_chain',
        _REJECT: 'fail_data' // track who failed, to know who should retry (and to respond to chain if it fails again)
      }
    },
    report_to_chain: {
      entry: ['send_report'],
      on: {
        _RESOLVE: '#idleState',
        _REJECT: 'fail_to_report'
      }
    },
    fail_to_connect: {
      on: {
        _RETRY: 'connect_to_encoders_and_hosters'
      }
    },
    fail_data: {
      on: {
        _RETRY: 'data_from_encoders_to_hosters'
      }
    },
    fail_to_report: {
      on: {
        _RETRY: 'report_to_chain'
      }
    },
  }
}
