// attest-state-machine/sub-performance-benchmark.js
module.exports = {
  initial:'p2plex_to_hosters',
  states: {
    p2plex_to_hosters: {
      entry: ['p2plex_connect_to_all_hosters'],
      on: {
        RESOLVE: 'join_attestors_swarm',
        REJECT: 'fail_to_connects'
      }
    },
    join_attestors_swarm: {
      entry: ['join_attestors_swarm'],
      on: {
        RESOLVE: 'compare_hoster_reports',
        REJECT: 'fail_to_join_swarm'
      }
    },
    compare_hoster_reports: {
      entry: ['compare_reports_with_other_attestors'],
      on: {
        RESOLVE: 'attestation',
        REJECT: 'fail_to_compare_reports'
      }
    },
    attestation: {
      entry: ['select_attestors', 'check_performance_for_selected_chunks'],
      on: {
        RESOLVE: 'performance_benchmark_report',
        REJECT: 'fail_to_attest'
      }
    },
    performance_benchmark_report: {
      entry: ['compare_reports_with_other_attestors'],
      exit: ['disconnect'],
      on: {
        RESOLVE: '#idleState',
        REJECT: 'fail_to_report'
      }
    },
    fail_to_connects: {
      on: {
        RETRY: 'p2plex_to_hosters'
      }
    },
    fail_to_join_swarm: {
      on: {
        RETRY: 'join_attestors_swarm'
      }
    },
    fail_to_compare_reports: {
      on: {
        RETRY: 'compare_hoster_reports'
      }
    },
    fail_to_attest: {
      on: {
        RETRY: 'attestation'
      }
    },
    fail_to_report: {
      on: {
        RETRY: 'performance_benchmark_report'
      }
    },
  }
}
