// attest-state-machine/sub-performance-benchmark-regular.js
module.exports = {
  initial:'join_attestors_swarm',
  states: {
    join_attestors_swarm: {
      entry: ['join_attestors_swarm'],
      on: {
        RESOLVE: 'idle',
        REJECT: 'fail_join_attestors_swarm'
      }
    },
    idle: {
      on: {
        PERFORMANCE_BENCHMARK_START: {
          target: 'performance_check'
        },
        PERFORMANCE_BENCHMARK_FINISHED: {
          target: 'performance_benchmark_finished'
        }
      },
    },
    performance_check: {
      entry: ['check_performance_for_selected_chunks'],
      on: {
        _RESOLVE: 'store_results',
        _REJECT: 'fail_performance_check'
      }
    },
    store_results: {
      entry: ['store_results_to_hyperlog'],
      on: {
        RESOLVE: 'idle',
        REJECT: 'fail_store_results'
      }
    },
    performance_benchmark_finished: {
      entry: ['disconnect'],
      on: {
        RESOLVE: '#idleState',
        REJECT: 'fail_performance_benchmark_finished'
      }
    },
    fail_join_attestors_swarm: {
      on: {
        RETRY: 'join_attestors_swarm'
      }
    },
    fail_performance_check: {
      on: {
        RETRY: 'performance_check'
      }
    },
    fail_store_results: {
      on: {
        RETRY: 'store_results'
      }
    },
    fail_performance_benchmark_finished: {
      on: {
        RETRY: 'performance_benchmark_finished'
      }
    },
  }
}
