// user-state-machine/index.js
const XState = require('xstate')
const { Machine, interpret, assign } = XState

module.exports = Machine({
  initial: 'idle',
  states: {
    idle: {
      on: {
        _ACCOUNT_CREATE: {
          target: 'account_create'
        },
        _ACCOUNT_DELETE: {
          target: 'account_delete'
        },
        _ACCOUNT_UPDATE: {
          target: 'account_update'
        }
      },
    },
    account_create: {
      entry: ['make_or_import_chain_keypair', 'init_nonce'],
      on: {
        _RESOLVE: 'roles_register',
        _REJECT: 'fail_account_create'
      }
    },
    roles_register: {
      entry: ['make_noise_keypair', 'init_storage', 'make_hyperlog'],
      on: {
        _RESOLVE: 'publish_to_chain',
        _REJECT: 'fail_roles_register'
      }
    },
    account_update: {
      entry: ['update'],
      on: {
        RESOLVE: 'publish_to_chain',
        REJECT: 'fail_account_update'
      }
    },
    account_delete: {
      entry: ['delete'],
      on: {
        RESOLVE: 'publish_to_chain',
        REJECT: 'fail_account_delete'
      }
    },
    publish_to_chain: {
      entry: ['make_or_import_noise_keypair', 'init_storage', 'make_hyperlog'],
      on: {
        RESOLVE: 'idle',
        REJECT: 'fail_publish_to_chain'
      }
    },
    fail_account_create: {
      on: {
        _RETRY: 'account_create'
      }
    },
    fail_account_update: {
      on: {
        _RETRY: 'account_update'
      }
    },
    fail_account_delete: {
      on: {
        _RETRY: 'account_delete'
      }
    },
    fail_roles_register: {
      on: {
        _RETRY: 'roles_register'
      }
    },
    fail_publish_to_chain: {
      on: {
        _RETRY: 'publish_to_chain'
      }
    },
  }
})
